crates/uv-auth/src/cache.rs

```diff
@@ -50,84 +23,56 @@ impl CredentialsCache {
-                trace!("Existing credentials include password, skipping cache");
-                // No need to look-up, we have a password already
-                return CheckResponse::Uncached(credentials);
-            }
-            trace!("Existing credentials missing password, checking cache");
-            let existing = store.get(&key);
-            existing
-                .cloned()
-                .map(CheckResponse::Cached)
-                .inspect(|_| trace!("Found cached credentials."))
-                .unwrap_or_else(|| {
-                    trace!("No credentials in cache, using existing credentials");
-                    CheckResponse::Uncached(credentials)
-                })
+    /// Return the credentials that should be used for a realm, if any.
+    pub(crate) fn get_realm(&self, netloc: NetLoc, username: Username) -> Option<Arc<Credentials>> {
+        let realms = self.realms.lock().unwrap();
+        let name = if let Some(username) = username.as_deref() {
+            format!("{username}@{netloc}")
         } else {
-            trace!("No credentials on request, checking cache...");
-            store
-                .get(&key)
-                .cloned()
-                .map(CheckResponse::Cached)
-                .inspect(|_| trace!("Found cached credentials."))
-                .unwrap_or_else(|| {
-                    trace!("No credentials in cache.");
-                    CheckResponse::None
-                })
-        }
+            netloc.to_string()
+        };
+        let key = (netloc, username);
+
+        realms
+            .get(&key)
+            .cloned()
+            .map(Some)
+            .inspect(|_| trace!("Found cached credentials for realm {name}"))
+            .unwrap_or_else(|| {
+                trace!("No credentials in cache for realm {name}");
+                None
+            })
     }
 
-    /// Update the cache with the given credentials if none exist.
-    pub(crate) fn set_default(&self, url: &Url, credentials: Arc<Credentials>) {
-        // Do not cache empty credentials
-        if credentials.is_empty() {
-            return;
-        }
-
-        // Insert an entry for requests including the username
-        if let Some(username) = credentials.username() {
-            let key = CredentialsCache::key(url, Some(username.to_string()));
-            if !self.contains_key(&key) {
-                self.insert_entry(key, credentials.clone());
+    /// Return the cached credentials for a URL, if any.
+    ///
+    /// The caller must not pass a URL with a username attached.
+    pub(crate) fn get_url(&self, url: &Url, username: Username) -> Option<Arc<Credentials>> {
+        debug_assert!(url.username().is_empty());
+        let urls = self.urls.lock().unwrap();
+        for ((cached_url, cached_username), credentials) in urls.iter() {
+            if cached_username == &username && url.as_str().starts_with(cached_url.as_str()) {
```

Comment:
`url.as_str().starts_with(cached_url.as_str())` feels off to me. (1) Why are those the right semantics? And (2) can we do this by matching on the _data_ rather than the string representation?

Reply:
What would happen if I made a request to:

`https://index.org/simple/botocore/?format=v4+application/json`

And it hits the netrc to grab the credentials. Then we cache the credentials under `https://index.org/simple/botocore/?format=v4+application/json`.

That path then returns a relative URL to an artifact, like `https://index.org/artifacts/foo-123.whl`. How would that latter URL get authenticated, since it wouldn't match here? Would we fetch the auth again? Would it go through the realm cache?

Reply:
We'd cache under `https://index.org/simple/botocore/` — we strip the query string when writing to the cache. But yeah the artifact would have to fall back to the realm cache since it's not a prefix.

The reason we use a prefix is a part of the RFC for caching credentials as well as a heuristic, e.g. credentials for `https://index.org/simple/botocore/` should be valid for `https://index.org/simple/botocore/foo/bar`.

The string representation of the URL is cached on the type so it's not expensive. It seems like the simplest way to check for a prefix-match. I could investigate further though?

Reply:
Note there is a bug here because `example.com/foo-2` would match `example.com/foo` but we should treat these as distinct resources e.g. `example.com/foo-2/` and `example.com/foo/`

Reply:
Resolved in the trie implementation